File: android/core/network/src/main/kotlin/xyz/klinik/network/LabApi.kt
package xyz.klinik.network

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
enum class LabFlag {
    LOW,
    NORMAL,
    HIGH,
    CRITICAL,
    ;

    val stringKey: String get() = "lab.flag.$name"
}

@Serializable
data class LabResult(
    val id: String,
    /** LOINC, where the printed name could be mapped. */
    val analyteCode: String? = null,
    val analyteName: String,
    /** Decimal on the wire, because a binary float cannot hold 0.1 exactly. */
    val value: String,
    val unit: String,
    val refLow: String? = null,
    val refHigh: String? = null,
    /** Null when the report carried no range — unclassified, not normal. */
    val flag: LabFlag? = null,
    val measuredAt: String,
    val ocrConfidence: String? = null,
    val verifiedAt: String? = null,
) {
    val referenceText: String?
        get() = when {
            refLow != null && refHigh != null -> "$refLow – $refHigh"
            refHigh != null -> "< $refHigh"
            refLow != null -> "> $refLow"
            else -> null
        }
}

/**
 * One report's worth of results, as the printed sheet had them.
 *
 * A lab report is read as a panel — the whole sheet from one blood draw — and
 * a screen that scattered those rows into per-analyte lists would make a
 * clinician reassemble on their own what the laboratory already grouped.
 */
@Serializable
data class LabPanel(
    /** When the sample was taken, not when it was uploaded. */
    val measuredAt: String,
    /** The report it came from. */
    val documentId: String? = null,
    val documentName: String? = null,
    /**
     * False when the row names a report that has no bytes behind it.
     *
     * The button to open it is not drawn in that case: a download that always
     * fails is worse than no download, because it looks like a broken app
     * rather than a missing file.
     */
    val documentAvailable: Boolean = false,
    /** Alphabetical by analyte name. */
    val results: List<LabResult> = emptyList(),
) {
    /** Whether anything on this sheet is outside its range. */
    val hasAbnormal: Boolean
        get() = results.any { it.flag == LabFlag.HIGH || it.flag == LabFlag.LOW }

    val hasCritical: Boolean get() = results.any { it.flag == LabFlag.CRITICAL }

    /**
     * Results with no range printed beside them.
     *
     * Unclassified rather than normal — the report carried no reference, and
     * drawing them as normal would be the client deciding something the
     * laboratory did not.
     */
    val unclassified: List<LabResult> get() = results.filter { it.flag == null }
}

@Serializable
data class LabReviewItem(
    val result: LabResult,
    /** The engine was unsure; one to look at first. */
    val needsAttention: Boolean,
    /** The printed name has no code yet. */
    val awaitingMapping: Boolean,
)

/**
 * The corrections a reviewer made. Everything is optional: confirming without
 * changing anything is the common case and must not require restating the row.
 */
@Serializable
data class LabCorrection(
    val analyteName: String? = null,
    val analyteCode: String? = null,
    val value: Double? = null,
    val unit: String? = null,
    val refLow: Double? = null,
    val refHigh: Double? = null,
)

@Serializable
data class TrendPoint(
    val measuredAt: String,
    val value: Double,
    val flag: LabFlag? = null,
    /** The range this particular result was measured against. */
    val refLow: Double? = null,
    val refHigh: Double? = null,
)

@Serializable
data class ReferenceBand(val low: Double? = null, val high: Double? = null)

@Serializable
data class AnalyteTrend(
    val analyteCode: String? = null,
    val analyteName: String,
    /** Series are split by unit: the same analyte in two units is two series. */
    val unit: String,
    val points: List<TrendPoint> = emptyList(),
    /**
     * Null when the points were measured against different ranges — drawing one
     * band across them would put results on the wrong side of a line they were
     * never compared to.
     */
    val reference: ReferenceBand? = null,
    val latestFlag: LabFlag? = null,
) {
    val id: String get() = "${analyteCode ?: analyteName}|$unit"
}

class LabApi(
    private val client: ApiClient,
    private val json: Json = ApiClient.defaultJson,
) {
    /** What OCR read and nobody has confirmed. Least certain first. */
    suspend fun pending(patientId: String): List<LabReviewItem> =
        decode(client.send(Endpoint(HttpMethod.GET, "patients/$patientId/lab-results/pending")))

    /** Confirmed results as per-analyte series, ready to chart. */
    suspend fun trends(subject: RecordSubject, since: String? = null): List<AnalyteTrend> =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.GET,
                    subject.base("lab-results/trends"),
                    query = since?.let { mapOf("since" to it) } ?: emptyMap(),
                ),
            ),
        )

    /**
     * Confirmed results grouped the way the laboratory printed them.
     *
     * Works for either subject: a patient reads their own panels through
     * `me/…`, a clinician reads a file through `patients/<id>/…`, and the
     * server decides what each may see.
     */
    suspend fun panels(subject: RecordSubject): List<LabPanel> =
        decode(client.send(Endpoint(HttpMethod.GET, subject.base("lab-results/panels"))))

    /** Confirmed values far enough outside their range to need attention now. */
    suspend fun critical(patientId: String): List<LabResult> =
        decode(client.send(Endpoint(HttpMethod.GET, "patients/$patientId/lab-results/critical")))

    /** Confirmed results only — what a trend may be drawn from. */
    suspend fun verified(patientId: String, analyteCode: String? = null): List<LabResult> =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.GET,
                    "patients/$patientId/lab-results",
                    query = analyteCode?.let { mapOf("analyteCode" to it) } ?: emptyMap(),
                ),
            ),
        )

    suspend fun verify(resultId: String, correction: LabCorrection): LabResult =
        decode(
            client.send(
                Endpoint(
                    HttpMethod.PATCH,
                    "lab-results/$resultId/verify",
                    body = json.encodeToString(LabCorrection.serializer(), correction),
                ),
            ),
        )

    /** For the table headings and page furniture OCR reads as values. */
    suspend fun discard(resultId: String) {
        client.send(Endpoint(HttpMethod.DELETE, "lab-results/$resultId"))
    }

    private inline fun <reified T> decode(body: String): T =
        runCatching { json.decodeFromString<T>(body) }
            .getOrElse { throw ApiError.Decoding(it.message ?: "unreadable response") }
}

File: ios/Sources/KlinikAPI/LabAPI.swift
import Foundation
import KlinikCore

public enum LabFlag: String, Decodable, Sendable, Equatable {
    case low = "LOW"
    case normal = "NORMAL"
    case high = "HIGH"
    case critical = "CRITICAL"

    public var localizedName: String { L10n.string("lab.flag.\(rawValue)") }
}

public struct LabResult: Decodable, Sendable, Equatable, Identifiable {
    public let id: String
    /// LOINC, where the printed name could be mapped.
    public let analyteCode: String?
    public let analyteName: String
    /// Decimal on the wire, because a binary float cannot hold 0.1 exactly.
    public let value: String
    public let unit: String
    public let refLow: String?
    public let refHigh: String?
    /// Null when the report carried no reference range — unclassified, not normal.
    public let flag: LabFlag?
    public let measuredAt: Date
    public let ocrConfidence: String?
    public let verifiedAt: Date?

    /**
     * Whether this one is outside its range.
     *
     * A result with no range is **not** normal — it is unclassified, and the
     * table says so rather than showing a tick nobody earned.
     */
    public var isOutOfRange: Bool {
        flag == .low || flag == .high || flag == .critical
    }

    public var referenceText: String? {
        switch (refLow, refHigh) {
        case let (low?, high?): return "\(low) – \(high)"
        case let (nil, high?): return "< \(high)"
        case let (low?, nil): return "> \(low)"
        default: return nil
        }
    }
}

/**
 * One lab report, as it was printed (spec M16).
 *
 * A trend answers "is this getting better"; a panel answers "what did the
 * blood test say" — every analyte at once, each beside its reference range.
 * The document travels with it so the reader can open the original: the table
 * is what OCR read, and the PDF is what the laboratory printed.
 */
public struct LabPanel: Decodable, Sendable, Equatable, Identifiable {
    public let measuredAt: Date
    public let documentId: String?
    public let documentName: String?

    /**
     * Whether there are bytes behind the report.
     *
     * A row can name one that has none. The reader is not offered a button
     * that opens an error — and the screen says the table is all there is,
     * rather than leaving them to wonder where the PDF went.
     *
     * Defaults to false when the server does not say, which is the safe
     * reading: no offer rather than a broken one.
     */
    public let documentAvailable: Bool

    public let results: [LabResult]

    /// Two reports on the same morning are two reports, so the document is
    /// part of what makes this one identifiable.
    public var id: String { "\(measuredAt.timeIntervalSince1970)|\(documentId ?? "")" }

    /// How many fall outside their reference range — the number worth a badge,
    /// and the reason somebody opens a report they were told was fine.
    public var abnormal: Int {
        results.filter(\.isOutOfRange).count
    }

    /// Present when there is a report to open, which is the only reason the
    /// button exists.
    public var canOpenReport: Bool { documentAvailable && documentId != nil }

    public init(
        measuredAt: Date,
        documentId: String? = nil,
        documentName: String? = nil,
        documentAvailable: Bool = false,
        results: [LabResult]
    ) {
        self.measuredAt = measuredAt
        self.documentId = documentId
        self.documentName = documentName
        self.documentAvailable = documentAvailable
        self.results = results
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        measuredAt = try container.decode(Date.self, forKey: .measuredAt)
        documentId = try container.decodeIfPresent(String.self, forKey: .documentId)
        documentName = try container.decodeIfPresent(String.self, forKey: .documentName)
        documentAvailable =
            try container.decodeIfPresent(Bool.self, forKey: .documentAvailable) ?? false
        results = try container.decode([LabResult].self, forKey: .results)
    }

    private enum CodingKeys: String, CodingKey {
        case measuredAt, documentId, documentName, documentAvailable, results
    }
}

public struct LabReviewItem: Decodable, Sendable, Equatable, Identifiable {
    public let result: LabResult
    /// The engine was unsure; this is one to look at first.
    public let needsAttention: Bool
    /// The printed name has no code yet.
    public let awaitingMapping: Bool

    public var id: String { result.id }
}

/// The corrections a reviewer made. Everything is optional: confirming without
/// changing anything is the common case and must not require restating the row.
public struct LabCorrection: Encodable, Sendable, Equatable {
    public var analyteName: String?
    public var analyteCode: String?
    public var value: Double?
    public var unit: String?
    public var refLow: Double?
    public var refHigh: Double?

    public init(
        analyteName: String? = nil,
        analyteCode: String? = nil,
        value: Double? = nil,
        unit: String? = nil,
        refLow: Double? = nil,
        refHigh: Double? = nil
    ) {
        self.analyteName = analyteName
        self.analyteCode = analyteCode
        self.value = value
        self.unit = unit
        self.refLow = refLow
        self.refHigh = refHigh
    }
}

public struct TrendPoint: Decodable, Sendable, Equatable, Identifiable {
    public let measuredAt: Date
    public let value: Double
    public let flag: LabFlag?
    /// The range this particular result was measured against.
    public let refLow: Double?
    public let refHigh: Double?

    public var id: Date { measuredAt }
}

public struct ReferenceBand: Decodable, Sendable, Equatable {
    public let low: Double?
    public let high: Double?
}

public struct AnalyteTrend: Decodable, Sendable, Equatable, Identifiable {
    public let analyteCode: String?
    public let analyteName: String
    /// Series are split by unit: the same analyte in two units is two series.
    public let unit: String
    public let points: [TrendPoint]
    /// Nil when the points were measured against different ranges — drawing one
    /// band across them would put results on the wrong side of a line they were
    /// never compared to.
    public let reference: ReferenceBand?
    public let latestFlag: LabFlag?

    public var id: String { "\(analyteCode ?? analyteName)|\(unit)" }
}

public struct LabAPI: Sendable {
    private let client: APIClient

    public init(client: APIClient) {
        self.client = client
    }

    /// What OCR read and nobody has confirmed. Least certain first.
    public func pending(patientId: String) async throws -> [LabReviewItem] {
        try await client.send(
            Endpoint(method: .get, path: "patients/\(patientId)/lab-results/pending"),
            as: [LabReviewItem].self
        )
    }

    /// Confirmed results as per-analyte series, ready to chart.
    public func trends(subject: RecordSubject, since: Date? = nil) async throws -> [AnalyteTrend] {
        var query: [String: String] = [:]
        if let since { query["since"] = ISO8601DateFormatter().string(from: since) }

        return try await client.send(
            Endpoint(method: .get, path: subject.base("lab-results/trends"), query: query),
            as: [AnalyteTrend].self
        )
    }

    /// Confirmed values far enough outside their range to need attention now.
    /// The reports, newest first. Confirmed results only — on both paths.
    public func panels(subject: RecordSubject) async throws -> [LabPanel] {
        try await client.send(
            Endpoint(method: .get, path: subject.base("lab-results/panels")),
            as: [LabPanel].self
        )
    }

    public func critical(patientId: String) async throws -> [LabResult] {
        try await client.send(
            Endpoint(method: .get, path: "patients/\(patientId)/lab-results/critical"),
            as: [LabResult].self
        )
    }

    /// Confirmed results only — what a trend may be drawn from.
    public func verify(resultId: String, correction: LabCorrection) async throws -> LabResult {
        try await client.send(
            Endpoint(
                method: .patch,
                path: "lab-results/\(resultId)/verify",
                body: try JSONEncoder.klinik.encode(correction)
            ),
            as: LabResult.self
        )
    }

    /// For the table headings and page furniture OCR reads as values.
    public func discard(resultId: String) async throws {
        try await client.send(Endpoint(method: .delete, path: "lab-results/\(resultId)"))
    }
}
